engines/e_loader_attic.c

```diff
@@ -783,8 +784,12 @@ static OSSL_STORE_INFO *try_decode_X509Certificate(const char *pem_name,
     if (cert == NULL)
         return NULL;
 
-    if ((d2i_X509_AUX(&cert, &blob, len)) != NULL
-        || (ignore_trusted && (d2i_X509(&cert, &blob, len)) != NULL)) {
+    create_cert = d2i_X509_AUX(&cert, &blob, len) != NULL;
+    if (!create_cert && ignore_trusted)
+        /* d2i_X509 will set cert but free the memory on error */
```

Comment:
Argh!  That's a bug that should be fixed *there*

Reply:
Quite a bit of code relies on the current behaviour.  Tests go crazy.

Reply:
``` diff
diff --git a/crypto/x509/x_x509.c b/crypto/x509/x_x509.c
index 529d701bbb..ea664f8214 100644
--- a/crypto/x509/x_x509.c
+++ b/crypto/x509/x_x509.c
@@ -134,6 +134,9 @@ X509 *d2i_X509(X509 **a, const unsigned char **in, long len)
             if (free_on_error)
                 X509_free(cert);
             cert = NULL;
+            /* Prevent *a from pointing at freed memory */
+            if (a != NULL)
+                *a = NULL;
         }
     }
     return cert;
```

Reply:
That's exactly the change I tried.  It broke things.

Reply:
> Quite a bit of code relies on the current behaviour. Tests go crazy.

Really?  I gotta check that out...

Reply:
(it could be that there are faults in the tests)

Reply:
A neater change is just: `*a = cert = NULL;`  `a` cannot be NULL at this point.

Reply:
Figure it out I think.  Fix coming soon.

Reply:
Pushed.  The fix turned out nice and simple.